sw/host/opentitantool/opentitantool_derive/src/lib.rs

```diff
@@ -0,0 +1,105 @@
+///
+/// Imagine that you have structs `World` and `People` which implement
+/// "Hello World" and "Hello People" commands.  You would create Newtype
+/// variants inside of a `Hello` enum and derive `CommandDispatch` to
+/// generate the dispatch layer for this interior node in the command hierarchy:
+///
+/// ```
+/// #[derive(StructOpt, CommandDispatch)]
+/// pub enum Hello {
+///     World(World),
+///     People(People),
+/// }
+///
+/// // The derived code is as follows:
+/// impl crate::command::CommandDispatch for Hello {
+///     fn run(
+///         &self,
+///         backend: &mut dyn opentitanlib::transport::Transport
+///     ) -> anyhow::Result<Option<Box<dyn erased_serde::Serialize>>> {
+///         match self {
+///             Hello::World(ref __field) => __field.run(backend),
+///             Hello::People(ref __field) => __field.run(backend),
+///         }
+///     }
+/// }
+/// ```
+#[proc_macro_derive(CommandDispatch)]
+#[proc_macro_error]
+pub fn derive_command_dispatch(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    match &input.data {
+        Data::Enum(e) => dispatch_enum(&input.ident, e).into(),
+        _ => abort!(
+            input.ident.span(),
+            "CommandDispatch is only implemented for enums"
+        ),
+    }
+}
+
+fn dispatch_enum(name: &Ident, e: &DataEnum) -> TokenStream {
+    let arms = e
+        .variants
+        .iter()
+        .map(|variant| dispatch_variant(name, variant))
+        .collect::<Vec<_>>();
+
+    // We wrap the derived code inside an anonymous const block to give the
+    // `extern crate` references a local namespace that wont pollute the
+    // global namespace.
+    quote! {
+        const _: () = {
+            extern crate opentitanlib;
+            extern crate anyhow;
+            extern crate erased_serde;
+
+            impl crate::command::CommandDispatch for #name {
+                fn run(
+                    &self,
+                    backend: &mut dyn opentitanlib::transport::Transport
+                ) -> anyhow::Result<Option<Box<dyn erased_serde::Serialize>>> {
+                    match self {
+                        #(#arms),*
+                    }
+                }
+            }
+        };
+    }
+}
+
+fn dispatch_variant(name: &Ident, variant: &Variant) -> TokenStream {
+    let ident = &variant.ident;
+    let unnamed_len = match &variant.fields {
+        Fields::Unnamed(u) => u.unnamed.len(),
+        _ => abort!(
+            variant.ident.span(),
+            "CommandDispatch is only implemented for Newtype variants"
+        ),
+    };
+    if unnamed_len != 1 {
+        abort!(
+            variant.ident.span(),
+            "CommandDispatch is only implemented for Newtype variants"
+        );
+    }
+    quote! {
+        #name::#ident(ref __field) =>
+            crate::command::CommandDispatch::run(__field, backend)
```

Comment:
Is `crate` right here? It doesn't work like `$crate` in macros, unfortunately.

Reply:
Per our 1:1 discussion, I've moved the proc macro and the `CommandDispatch` trait definition into `opentitanlib` so that they can be referred to unambiguously.